### PR TITLE
Filters able to raise alerts themselves

### LINKED ISSUE #104: [IDEA] Remove flogs for an "AlertManager" in the Core
**What is your feature about ?**
Removing the flogs filter because it does not comply with darwin philosophy.

Creating an "AlertManager" in the Core that is able to handle alert file, redis and maybe more.

**Additional context**
The "AlertManager" could use a list of output interfaces _(IAlertRepository)_.

### LINKED ISSUE #136: [IDEA] Do not alert on error (certitude = 101)
**What is your feature about ?**

Not logging errors on entries as alerts.
Errors are shown by the certitude at 101.

### PR DESCRIPTION
**Pull request by HugoSoszynski**
# :sparkles: Pull Request Template
:bangbang: Once all the **checklist** is **done** you have to:
  * **stash merge** this pull request
  * **delete** the corresponding **branch**
  * **close** the associated **issue**

## :page_with_curl: Type of change

**Breaking change**: fix or feature that would cause existing functionality to not work as expected.

## :bulb: Related Issue(s)

- Resolve #104 
- Resolve #136 

## :black_nib: Description

**BREAKING**: The filters no longer launch alert on 101 code (error code).

**CAUTION**: This PR implies new configuration fields for the filters. However those are not mandatory (see filters documentation).

Adding an AlertManager to the Core to enable every filter to raise an alert.

The alerts can be raised in : 
- File
- Redis list
- Redis channel

Alert file rotation is handled catching the SIGHUP or SIGUSR1 signal.

### Core
The new alerting configuration fields are:
 - 'log_file_path': path to the log file dedicated to alerts
 - 'redis_socket_path': path to the redis UNIX socket to send alert to
 - 'alert_redis_list_name': name of the alert list in redis
 - 'alert_redis_channel_name': name of redis channel to publish alert to

If the alert configuration is missing or wrong, it is not applied.
If the alert configuration is partialy correct, apply as much as possible.
In both cases a warning is logged.

DARWIN_RAISE_ALERT(str) macro makes filter code easier to read. (AlertManager.hpp)

### CMake
- Added the Redis dependencies to the CMakeLists.txt because it will soon be part of all the filters through the Alert Manager.
- Added the required toolkit parts to the core sources.
- Added DARWIN_LIBRARIES var to remove duplicata and avoid core libraries mistake.

### FileManager
- Only accepts const strings as no modification should occur.
- Rewrote parts of the FileManager to be more C++ friendly.
- Force reopen is now possible with our FiLeManager.
- File opening / reopening is now done under mutex lock to prevent writing while opening / closing the file.

### Filters
- Filters are now using the AlertManager to raise alert.
- Added ftest to enable easier testing of the core functions. This filter is not compiled by default but is mandatory for tests.

### Tests
- Added the alert manager tests (around 75 tests so no listing here)
- Updated TAnomaly tests to check for proper handling of both legacy and new configuration file.

### FreeBSD (version)
- Redis 4.0.14_1
- Boost 1.71.0
- clang++ 8.0.1
- CMake 3.15.4
- Python 3.6

### Ubuntu (version)
- Redis 5.0.3
- Boost 1.71.0
- g++ 8.3.0
- CMake 3.13.14
- Python 3.7.3
- Valgrind 3.14.0

## :heavy_check_mark: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] _(ftest)_ I have added corresponding page to the documentation
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

</br>

- [x] :raising_hand: **I certify on my honor that all the information provided is true, and I've done all I can to deliver a high quality code**

### CODE REVIEW

**Comment by HugoSoszynski** [comment]
As the log rotate mechanism was put to my attention, this will require a dev and associated tests.

File: tests/core/alert.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+# wfile & socket & wlist & channel
+# wfile & socket & list & wchannel
+# wfile & wsocket & wlist & channel
+# wfile & socket & wlist & wchannel
+# wfile & wsocket & list & wchannel
+# wfile & wsocket & wlist & wchannel
+def run():
+    tests = [
+        # {"test_name": "no_config_file", "conf": None, "log": "this is log 1", "expected_file": False, "expected_list": False, "expected_channel": False},
+        # {"test_name": "config_file_empty", "conf": "", "log": "this is log 2", "expected_file": False, "expected_list": False, "expected_channel": False},
```
Comment (frikilax) [inline tests/core/alert.py]: dead code

File: tests/core/alert.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        if not file_content and not expected:
+            return True
+        elif not file_content and expected:
+            if file_content is None:
+                logging.error("Was expecting alert in log file, file doesn't exist")
+            else:
+                logging.error("Was expecting alert in log file, file is empty")
+            return False
+        elif file_content and not expected:
+            logging.error("Expected log file to be empty or non extisting but got content")
```
Comment (frikilax) [inline tests/core/alert.py]:
```suggestion
            logging.error("Expected log file to be empty or non extisting but got {}".parse(file_content))
```

File: tests/core/alert.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    if not file_content and not expected:
+        return True
+    elif not file_content and expected:
+        if file_content is None:
+            logging.error("Was expecting alert in log file, file doesn't exist")
+        else:
+            logging.error("Was expecting alert in log file, file is empty")
+        return False
+    elif file_content and not expected:
+        logging.error("Expected log file to be empty or non extisting but got content")
```
Comment (frikilax) [inline tests/core/alert.py]:
```suggestion
        logging.error("Expected log file to be empty or non existing but got content")
```

File: toolkit/FileManager.hpp [hunk 4]
```diff
@@ -18,23 +18,29 @@ namespace darwin {
             /// FileManager constructor
             /// \param file the file's path to manage
             /// \param append if we append data to the file
-            explicit FileManager(std::string& file, bool app=true);
+            explicit FileManager(const std::string& file, bool app=true);
             ~FileManager();
 
             /// Open the file
+            /// \param froce_reopen If true, close the file and reopen it no matter what
```
Comment (frikilax) [inline toolkit/FileManager.hpp]:
```suggestion
            /// \param force_reopen If true, close the file and reopen it no matter what
```

File: tests/core/alert.py (hunk 1 above)
Comment (HugoSoszynski) [reply tests/core/alert.py]: Done in 87857a3

File: samples/base/AlertManager.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    bool AlertManager::LoadRedisConfig(const rapidjson::Document& configuration) {
+        DARWIN_LOGGER;
+
+        if (!configuration["redis_socket_path"].IsString() || configuration["redis_socket_path"].GetStringLength() <= 0) {
+            DARWIN_LOG_WARNING("AlertManager:: 'redis_socket_path' needs to be a non-empty string. Ignoring REDIS configuration...");
+            _redis = false; // Error configuring redis, disabling...
+            return false;
+        }
+        std::string redis_socket_path = configuration["redis_socket_path"].GetString();
```
[the remark this answers is not in the document]
Comment (HugoSoszynski) [reply samples/base/AlertManager.cpp]: Done in 208ed03

File: samples/base/AlertManager.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        unsigned int retry = RETRY;
+        bool fail;
+
+        fail = !(_log_file->Write(str + '\n'));
+        while(retry and fail){
+            DARWIN_LOG_INFO("AlertManager::WriteLogs:: Error when writing in log file, "
+                            "will retry " + std::to_string(retry) + " times");
+            fail = !(_log_file->Write(str + '\n'));
+            retry--;
+        }
```
[the remark this answers is not in the document]
Comment (HugoSoszynski) [reply samples/base/AlertManager.cpp]: Done in 208ed03

File: tests/core/alert.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+# wfile & wsocket & wchannel
+#
+# wfile & socket & list & channel
+# wfile & wsocket & list & channel
+# wfile & socket & wlist & channel
+# wfile & socket & list & wchannel
+# wfile & wsocket & wlist & channel
+# wfile & socket & wlist & wchannel
+# wfile & wsocket & list & wchannel
+# wfile & wsocket & wlist & wchannel
```
[the remark this answers is not in the document]
Comment (HugoSoszynski) [reply tests/core/alert.py]: Done in 208ed03

File: tests/core/alert.py (hunk 2 above)
Comment (HugoSoszynski) [reply tests/core/alert.py]: Done in 208ed03

**Pull request merged**